### PR TITLE
[api/frontend] Status Template creation, deletion, edition (#1943)

### LINKED ISSUE #1943: Ability to create additional custom workflow status names straight from the UI if possible.
## Use case
Ability to create additional custom workflow status names straight from the UI if possible. 
Screen Shot 2022-03-04 at 11 46 12 AM

### PR DESCRIPTION
**Pull request by Archidoit**
### Proposed changes

*
*

### Related issues

*
*

### Checklist

- [ ] I consider the submitted work as finished
- [ ] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Kedae** [comment]
Couldn't there be a way that when we create a new Status template from the workflow edition drawer, it is automatically selected ?

**Comment by Kedae** [comment]
Workflows don't seem to be working for Actors, and may be some other objects.

**Comment by Kedae** [comment]
We have a mismatch of color handling: on worklows a status template without color is "grey" and in the creation of that status template we were given a random color in the list.
Same in the detail of a object where the status is used, first time -> random color then a white/grey color

File: opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import React, { useState } from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js]: This file should be .tsx

File: opencti-platform/opencti-front/src/private/components/settings/StatusTemplatesQuery.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { graphql } from 'react-relay';
+
+// eslint-disable-next-line import/prefer-default-export
+export const statusTemplatesSearchQuery = graphql`
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/StatusTemplatesQuery.js]: Why the need of a new file for this ?

File: opencti-platform/opencti-front/src/private/components/settings/Workflow.js [hunk 3]
```diff
@@ -80,6 +86,8 @@ class Sessions extends Component {
             return <div />;
           }}
         />
+        <WorkflowsStatusesMenu />
+        {view === 'lines' ? this.renderLines(paginationOptions) : ''}
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/Workflow.js]: We try as much as possible to drop this 'view === 'lines'' stuff as it is useless

File: opencti-platform/opencti-front/src/private/components/settings/Workflow.js [hunk 4]
```diff
@@ -57,8 +58,13 @@ class Sessions extends Component {
   }
 
   render() {
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/Workflow.js]: Could you try turning this into a pure function typescript component ?

File: opencti-platform/opencti-front/src/private/components/settings/Workflow.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ import {
   buildViewParamsFromUrlAndStorage,
   saveViewParameters,
 } from '../../../utils/ListParameters';
+import WorkflowsStatusesMenu from './workflow/WorkflowsStatusesMenu';
 
 const Transition = React.forwardRef((props, ref) => (
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/Workflow.js]: @SamuelHassine what is that ?

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusCreation.js [hunk 6]
```diff
@@ -1,24 +1,21 @@
-import React, { Component } from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusCreation.js:1]: Should be in .tsx

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusEdition.js [hunk 7]
```diff
@@ -1,58 +1,17 @@
-import React, { Component } from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusEdition.js:1]: Same

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusEdition.js [hunk 8]
```diff
@@ -69,129 +28,106 @@ const statusMutationFieldPatch = graphql`
...
+    template: {
+      label: data.template.name,
+      value: data.template.id,
+      color: data.template.color,
+    },
+    order: data.order,
+  };
+
+  const handleSubmitStatusTemplate = (values, { setSubmitting }) => {
+    commitMutation({
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusEdition.js]: Can you try using useMutation instead ?

**Comment by Archidoit** [comment]
> Workflows don't seem to be working for Actors, and may be some other objects.

--> delete the corresponding workflows statuses when a status template is deleted

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateCreation.js [hunk 9]
```diff
@@ -0,0 +1,297 @@
+import React, { useState } from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateCreation.js]: Same for typescript and useMutation

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateEdition.js [hunk 10]
```diff
@@ -0,0 +1,160 @@
+import React from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateEdition.js]: Same

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateEdition.js [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+  const handleSubmitField = (name, value) => {
+    statusTemplateValidation(t)
+      .validateAt(name, { [name]: value })
+      .then(() => {
+        commitMutation({
+          mutation: statusTemplateMutationFieldPatch,
+          variables: {
+            id: data.id,
+            input: { key: name, value: value || '' },
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateEdition.js]: Value should be able to be undefined or null here

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateEdition.js [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        <div className="clearfix" />
+      </div>
+      <div className={classes.container}>
+        <Formik
+          enableReinitialize={true}
+          initialValues={initialValues}
+          validationSchema={statusTemplateValidation(t)}
+        >
+          {() => (
+            <Form style={{ margin: '20px 0 20px 0' }}>
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateEdition.js]: Can you add a constant to the fieldUtils you created ?

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateLine.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
+import React from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateLine.js]: Same

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateLineDummy.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
+import React from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateLineDummy.js]: Same

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateLineDummy.js [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        <Skeleton
+          animation="wave"
+          variant="circular"
+          width={30}
+          height={30}
+        />
+      </ListItemIcon>
+      <ListItemText
+        primary={
+          <div>
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateLineDummy.js]: You should iterate over dataColumns

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatePopover.js [hunk 16]
```diff
@@ -0,0 +1,184 @@
+import React, { useState } from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatePopover.js]: Same

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplates.js [hunk 17]
```diff
@@ -0,0 +1,86 @@
+import React from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplates.js]: Same

**Comment by Archidoit** [comment]
> We have a mismatch of color handling: on worklows a status template without color is "grey" and in the creation of that status template we were given a random color in the list. Same in the detail of a object where the status is used, first time -> random color then a white/grey color

--> have a rule for invalid color string inputs

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatesLines.js [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+const nbOfRowsToLoad = 50;
+
+const StatusTemplatesLines = ({ relay, paginationOptions, dataColumns, data, initialLoading }) => {
+  return (
+    <ListLinesContent
+      initialLoading={initialLoading}
+      loadMore={relay.loadMore}
+      hasMore={relay.hasMore}
+      isLoading={relay.isLoading}
+      dataList={pathOr([], ['statusTemplates', 'edges'], data)}
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatesLines.js]: dataList={data?.statusTemplates?.edges ?? []}

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatesLines.js [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+      globalCount={pathOr(
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatesLines.js]: Same here

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatesLines.js [hunk 20]
```diff
@@ -0,0 +1,106 @@
+import React from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplatesLines.js]: Same

File: opencti-platform/opencti-front/src/private/components/settings/workflow/WorkflowsStatusesMenu.js [hunk 21]
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/WorkflowsStatusesMenu.js]: Same

File: opencti-platform/opencti-front/src/utils/Edition.js [hunk 22]
```diff
@@ -3,9 +3,7 @@ import * as R from 'ramda';
 export const convertStatus = (t, element) => (R.pathOr(null, ['status', 'template', 'name'], element) === null
   ? ''
   : {
-    label: t(
-      `status_${R.pathOr(null, ['status', 'template', 'name'], element)}`,
-    ),
+    label: R.pathOr(null, ['status', 'template', 'name'], element),
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/utils/Edition.js]: element?.status?.template?.name ?? null

**Comment by SamuelHassine** [comment]
Some issues to solve before merging this PR:

In the status templates list, color column is not aligned (certainly due to an incorrect configuration in the header).

image

When creating a new Status Template on-the-fly during the modification of an existing workflow, the name field is not correctly filled:

image

Also, some translations are missing, all related to the management of status templates like:

* Do you want to delete this status template?
* Create a status template
* Update a status template
* etc.

Last thing, can you please check why the status are not available in the report edition form?

image

File: opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+
+  const searchStatusTemplates = (event) => {
+    setStatusTemplateInput(event && event.target.value !== 0 ? event.target.value : '');
+    fetchQuery(statusTemplatesSearchQuery, {
+      search: event && event.target.value !== 0 ? event.target.value : '',
+    })
+      .toPromise()
+      .then((data) => {
+        const NewStatusTemplates = pipe(
+          pathOr([], ['statusTemplates', 'edges']),
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js]: Can we remove ramda by using javascript functions like statusTemplates?.edges?.map() ?

File: opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+          map((n) => ({
+            label: n.node.name,
+            value: n.node.id,
+            color: n.node.color,
+          })),
+        )(data);
+        setStatusTemplates(union(statusTemplates, NewStatusTemplates));
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js]: It's possible to use the concat() function

File: opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+      });
+  };
+
+  return (
+    <div>
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/common/form/StatusTemplateField.js]: Is this IVD necessary?

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusCreation.js [hunk 26]
```diff
@@ -87,152 +84,123 @@ const statusCreationMutation = graphql`
...
+    handleClose();
+  };
 
-StatusCreation.propTypes = {
-  display: PropTypes.bool,
-  classes: PropTypes.object,
-  theme: PropTypes.object,
-  t: PropTypes.func,
+  return (
+    <div style={{ display: display ? 'block' : 'none' }}>
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusCreation.js]: @Kedae It's not better to use the condition direclty on the parent/calling component to show or not this component ?

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateCreation.js [hunk 27]
```diff
@@ -0,0 +1,297 @@
...
+    handleClose();
+  };
+
+  const onResetContextual = () => {
+    handleCloseContextual();
+  };
+
+  const renderClassic = () => {
+    return (
+      <div>
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/workflow/StatusTemplateCreation.js]: IS this DIV necessary ?

File: opencti-platform/opencti-front/src/utils/Localization.js [hunk 28]
```diff
@@ -4520,12 +4508,6 @@ const i18n = {
       'infrastructure-operator': 'インフラストラクチャ運用者',
       'malware-author': 'マルウェア作成者',
       sponsor: 'スポンサー',
-      status_NEW: '新規',
-      status_IN_PROGRESS: '進行中',
-      status_ANALYZED: '解析完了',
-      status_CLOSED: '終了済',
-      status_PENDING: '保留',
-      status_TO_BE_QUALIFIED: '取り掛かる',
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/utils/Localization.js:4639]: Just for my knowledge, we use Google Translate for these languages ?

Comment (SamuelHassine) [reply opencti-platform/opencti-front/src/utils/Localization.js:4639]: Yes :) Then we will let native people to fix it if necessary.

File: opencti-platform/opencti-front/src/private/components/settings/workflow/StatusCreation.js (hunk 26 above)
Comment (Kedae) [reply opencti-platform/opencti-front/src/private/components/settings/workflow/StatusCreation.js]: Absolutely, I've missed that.

**Review by richard-julien — APPROVED** [review]

**Pull request merged**